java8/src/main/java/org/jetbrains/annotations/BlockingExecutor.java

```diff
@@ -0,0 +1,50 @@
+/*
+ * Copyright 2000-2021 JetBrains s.r.o.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.jetbrains.annotations;
+
+import java.lang.annotation.*;
+
+/**
+ * Indicates that the annotated executor (e.g. coroutine dispatcher, scheduler, etc.)
+ * allows blocking methods execution.
+ *
+ * If a given context does not allow blocking calls, {@link NonBlockingExecutor} should be used.
+ *
+ * Example:
+ * <pre>
+ * {@code
+ *  class ExampleService {
+ *      @BlockingContext
+ *      val dispatcher: CoroutineContext
+ *          get() { ... }
+ *
+ *      suspend fun foo() {
+ *          val result = withContext(dispatcher) {
+ *              blockingBuzz() // no IDE warning
+ *          }
+ *      }
+ *
+ *      @Blocking fun blockingBuzz() { ... }
+ *  }
+ * }
+ * </pre>
+ */
+@Documented
+@Retention(RetentionPolicy.CLASS)
+@Target({ElementType.TYPE, ElementType.FIELD, ElementType.METHOD, ElementType.PARAMETER, ElementType.TYPE_USE})
```

Comment:
What `ElementType.TYPE` here means? What does it mean if we annotate the type?

Also, probably it would be enough to keep TYPE_USE target only and remove everything else, just annotating via `val dispatcher: @BlockingContext CoroutineContext`? Using TYPE_USE and non-type-use targets at the same time produces ambiguity in Java code and we try to avoid this in new annotations

Reply:
`ElementType.TYPE` - If we have a custom dispatcher, we can annotate the corresponding class.

As for `TYPE_USE` only - we've discussed it, it is not suitable because in kotlin types are very frequently omitted (and in the example I've explicitly added type info just to make sure that reader will understand what exactly the `dispatcher` property is)

Reply:
@chashnikov I think we're ok with TYPE and TYPE_USE targets.

Reply:
On the other hand, annotating type with type_use-only annotation is legal in Java. E.g., this code is ok:

```java
import java.lang.annotation.ElementType;
import java.lang.annotation.Target;

@Foo
public class Main {
}
@Target(ElementType.TYPE_USE)
@interface Foo {}
```

Not sure about Kotlin. If it's ok in Kotlin as well, then keeping TYPE looks redundant.

Reply:
In Kotlin it is illegal to annotate type with annotation having only TYPE_USE target, so combining them two here looks reasonable for me